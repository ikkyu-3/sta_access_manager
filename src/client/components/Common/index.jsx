// @flow
import type { CommonProps } from "@types";

import * as React from "react";
import SingleButton from "@components/SingleButton";
import styles from "./style.css";

class Common extends React.Component<CommonProps> {
  componentDidMount() {
    this.timeoutId = setTimeout(() => {
      this.props.autoPageTransition();
    }, 5000);
  }

  componentWillUnmount() {
    clearTimeout(this.timeoutId);
  }

  timeoutId: TimeoutID;

  render(): React.Node {
    const {
      type,
      text,
      subText,
      buttonText,
      buttonClass,
      onButtonClick
    } = this.props;
    const { textArea, main, date, sub } = styles;

    const dateTextNode =
      type === "date" ? (
        <div className={date}>{new Date().toLocaleString()}</div>
      ) : null;
    const subTextNode = subText ? <div className={sub}>{subText}</div> : null;

    return (
      <div>
        <div className={textArea}>
          <div className={main}>{text}</div>
          {dateTextNode}
          {subTextNode}
        </div>
        <SingleButton
          text={buttonText}
          className={buttonClass}
          onButtonClick={onButtonClick}
        />
      </div>
    );
  }
}
export default Common;
