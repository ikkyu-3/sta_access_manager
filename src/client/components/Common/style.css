@charset "utf-8";
.textArea {
  position: relative;
  height: calc(100% - 70px);
}

.main {
  display: flex;
  height: 60%;
  justify-content: center;
  align-items: flex-end;
  font-size: 36px;
  font-weight: bold;
}

.date {
  display: flex;
  height: 40%;
  justify-content: center;
  font-size: 20px;
  padding-top: 10px;
  box-sizing: border-box;
}

.sub {
  position: absolute;
  right: 10px;
  bottom: 10px;
  font-size: 24px;
}
