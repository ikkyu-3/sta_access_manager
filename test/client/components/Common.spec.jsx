import React from "react";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import Common from "@components/Common";

// Dateクラスをmock化
jest
  .spyOn(Date.prototype, "toLocaleString")
  .mockReturnValue("1970/01/01 00:00:00");

const setTimeoutSpy = jest
  .spyOn(window, "setTimeout")
  .mockImplementation(() => 999);

const clearTimeoutSpy = jest
  .spyOn(window, "clearTimeout")
  .mockImplementation(() => {});

const fn = () => {};
const factory = (values = {}) => {
  const { type, subText } = values;
  return shallow(
    <Common
      type={type}
      text="テキスト"
      subText={subText}
      buttonText="ボタン"
      buttonClass="default"
      onButtonClick={fn}
      autoPageTransition={fn}
    />
  );
};

describe("Common.jsxのテスト", () => {
  describe("スナップショット", () => {
    it("正しいレンダリング", () => {
      const tree = renderer.create(
        <Common
          type="date"
          text="テキスト"
          subText="サブテキスト"
          buttonText="ボタン"
          buttonClass="default"
          onButtonClick={fn}
          autoPageTransition={fn}
        />
      );
      expect(tree).toMatchSnapshot();
    });
  });

  describe("コンポーネントのテスト", () => {
    let defaultCommon;

    beforeAll(() => {
      defaultCommon = factory();
    });

    it("componentDidMountは、setTimeoutを実行する", () => {
      const [first, second] = setTimeoutSpy.mock.calls[0];
      expect(first).toBeInstanceOf(Function);
      expect(second).toBe(5000);
      expect(defaultCommon.instance().timeoutId).toBe(999);
    });

    it("componentWillUnmountは、clearTimeoutを実行する", () => {
      defaultCommon.instance().componentWillUnmount();
      const [first] = clearTimeoutSpy.mock.calls[0];
      expect(first).toBe(defaultCommon.instance().timeoutId);
    });

    it("typeがdateの時、時刻が表示されている", () => {
      const common = factory({ type: "date" });
      expect(common.find(".date").text()).toBe("1970/01/01 00:00:00");
    });

    it("typeが指定されていない時、時刻は表示されていない", () => {
      expect(defaultCommon.find(".date").exists()).toBeFalsy();
    });

    it("typeにdate以外が指定されている時、時刻は表示されていない", () => {
      const common = factory({ type: "hoge" });
      expect(common.find(".date").exists()).toBeFalsy();
    });

    it("textに渡した値が表示されている", () => {
      expect(defaultCommon.find(".main").text()).toBe("テキスト");
    });

    it("subTextが指定されている時、subTextに渡した値が表示されている", () => {
      const common = factory({ subText: "サブテキスト" });
      expect(common.exists(".sub")).toBeTruthy();
      expect(common.find(".sub").text()).toBe("サブテキスト");
    });

    it("subTextが指定されていない場合、表示されていない", () => {
      expect(defaultCommon.find(".sub").exists()).toBeFalsy();
    });

    it("SingleButtonコンポーネントに、buttonText, buttonClass, onButtonClickに指定した値が正常に渡されている", () => {
      const singleButton = factory()
        .find("SingleButton")
        .props();

      expect(singleButton.text).toBe("ボタン");
      expect(singleButton.className).toBe("default");
      expect(singleButton.onButtonClick).toBe(fn);
    });
  });
});
