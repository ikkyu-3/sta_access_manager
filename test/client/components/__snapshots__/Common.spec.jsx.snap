// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Common.jsxのテスト スナップショット 正しいレンダリング 1`] = `
<div>
  <div
    className="textArea"
  >
    <div
      className="main"
    >
      テキスト
    </div>
    <div
      className="date"
    >
      1970/01/01 00:00:00
    </div>
    <div
      className="sub"
    >
      サブテキスト
    </div>
  </div>
  <div
    className="singleButton"
  >
    <button
      className="default"
      onClick={[Function]}
    >
      ボタン
    </button>
  </div>
</div>
`;
